core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1935,4 +2040,125 @@ object Log {
   private def isLogFile(file: File): Boolean =
     file.getPath.endsWith(LogFileSuffix)
 
+  @throws(classOf[IndexOffsetOverflowException])
+  private[log] def maybeRetryOnOffsetOverflow[T](maxTries: Int = 1)(fn: => T): T = {
```

Comment:
Would it make sense to move the splitting of the segment out of `recoverSegment` and into this function? Then it is clear how the overflow offset is being remediated, which will make the retry logic clearer.

Also, I think the condition for retrying should be based on whether we have made progress, not on a preset number of retries. If it is a swap file, then we will not make progress which has an overflow, then no progress will be made and we can rethrow. Otherwise, is there any reason we wouldn't make progress?

Reply:
Hmm, this is a higher level retry method that typically does not deal with individual segments. The expected workflow is for this to be used by a higher level operation that needs to be retried when a segment-specific method underneath ran into offset overflow. The segment-specific method would then split the segment and propagate the error up to the place where the entire operation can be retried safely, factoring in the new segments that were added (and the old segment that was deleted).

The contract is kind of co-operative in that sense - methods lower down the stack split the segment and expect that a method higher up catches the exception and retries. I think it is somewhat hard to assert here that the lower level method ended up actually splitting the segment. Let me know if you have any ideas on how we could do that.

Also, swap file recovery does not use this method.

Reply:
> The contract is kind of co-operative in that sense - methods lower down the stack split the segment and expect that a method higher up catches the exception and retries. I think it is somewhat hard to assert here that the method lower down the stack ended up actually splitting the segment. Let me know if you have any ideas on how we could do that.

Yeah, this is exactly my point. It's suboptimal for the retry logic to be decoupled from whatever condition is enabling the retry. The caller has to have some understanding of the retry mechanism anyway, as we do here. What if we add the `LogSegment` that has the overflow as a field to the `IndexOffsetOverflowException` exception? That would allow us to do the splitting here and take it out of the lower level logic.